src/liboslexec/batched_llvm_gen.cpp

```diff
@@ -7088,10 +7088,43 @@ llvm_gen_keyword_fill(BatchedBackendLLVM& rop, Opcode& op,
             if (equivalent(p.type, ValueType) && !strcmp(key->c_str(), p.key)) {
                 // store data
                 OSL_DASSERT(p.offset + p.field_size <= clentry->struct_size);
-                llvm::Value* dst = rop.ll.offset_ptr(mem_void_ptr, p.offset);
-                llvm::Value* src = rop.llvm_void_ptr(Value);
-                rop.ll.op_memcpy(dst, src, (int)p.type.size(),
-                                 4 /* use 4 byte alignment for now */);
+
+                bool arg_is_uniform = Value.is_uniform();
+
+                TypeDesc simpletype(Value.typespec().simpletype());
+                int num_elements   = simpletype.numelements();
+                int num_components = simpletype.aggregate;
+
+                llvm::Value* dest_base = rop.ll.offset_ptr(mem_void_ptr, p.offset);
+                dest_base              = rop.llvm_ptr_cast(dest_base, p.type);
+
+                for (int a = 0; a < num_elements; ++a) {
+                    llvm::Value* arrind = simpletype.arraylen ? rop.ll.constant(a)
+                                                               : NULL;
+                    for (int c = 0; c < num_components; c++) {
+                        llvm::Value* dest_elem;
+                        if (num_components > 1)
```

Comment:
What happens for an array of of objects with 1 compoenent?
Looks like maybe the else case is taken which disregards the "a" which is the element index.
Does the else case need to incorporate the arrayind or "a"

Reply:
Hmm, that does look like a bug to me too ... which quite possibly means it is also a bug on line 7243 where I copied the code from?

Not that I want to make it harder to get this quick fix I need over the line, but I do wonder if we should add a closure with more types of parameters to testshade so that it's possible to actually write testcases for this code.

Reply:
This same comment might apply to the copy loop in llvm_gen_closure of batched_llvm_gen.cpp

Reply:
This started out as a quick fix to a bug that was holding me up ... modifying testshade seems like a somewhat larger change - adding new closure parameters to testshade feels a bit more involved, since they would potentially be visible to other users of OSL.  Does someone want to decide what this should look like?  Should we add some keyword parameters to `debug()`?  Oh, wait, looking at the list of macros in genclosure.h, it looks like keyword array parameters aren't currently supported anyway?  So potentially I could drop the iteration code and use a simple memcpy - though the bug would still potentially exist in the non-keyword array parameter case that already exists.  Should we add a special `test` closure with a parameter of every type in genclosure.h?

Reply:
I agree, that looks like a bug.  I think the two main purposes of testshade are to provide an example of how to integrate the library and to allow for testing the important features of the library.  A special 'test' closure which exercises the types in genclosure.h would be something that accomplishes both, so sounds like a good idea to me -- it'll keep me from accidentally checking things like this in that work for our renderer which doesn't make heavy use of closures but break those that depend on them heavily.

Reply:
I think you can still get this fix over the fence without everything else, maybe remove the array index stuff if you know its not needed, replace it with an assert?  You have added a test case and modified source to allow it to pass without breaking any other testcases.  We just noticed some other issues along the way.  Maybe drop a "todo" on the line 7243 to add testcase and fix bug.